File: fpc_hosted/aarch64/test/sieve1.c
#include <stdio.h>

#define  SIZE 1000000
#define  ITERMAX 1000

int flags[SIZE + 1];

int i, prime, k, count, iter;

int main() {

  for (iter = 0; iter < ITERMAX; iter++) {
    count = 0;
    for (i = 0; i <= SIZE; i++) { 
       flags[i] = 1;
    }
    for (i = 0;  i <= SIZE; i++) {
      if (flags[i]) {
        prime = i + i + 3;
        k = i + prime;
        while (k <= SIZE) {
          flags[k] = 0;
          k = k + prime;
        }
        count = count + 1;
      }
    }
  }
  printf("count: %d\n", count);
  return 0;
}

File: fpc_hosted/aarch64/test/sieve1.pas
{$noindexcheck,norangecheck,nopointercheck,nostackcheck}
{$nowalkback}

{       Sieve of Eratosthenes benchmark      }
 
program sieve;
const
  SIZE = 1000;
  ITERMAX = 1000000;

var
  flags : array [0..SIZE] of boolean;
  i, prime, k, count, iter : integer;
 
%include 'testlib';

procedure foo; begin end;

begin
  for iter := 1 to ITERMAX do begin
    count := 0;
    for i := 0 to SIZE do 
       flags[i] := true;
    for i := 0 to SIZE do
      if flags[i] then begin
        prime := i + i + 3;
{
foo;
}
        k := i + prime;
        while k <= SIZE do begin
          flags[k] := false;
          k := k + prime;
        end;
        count := count + 1;
      end;
  end;
  putintln(count);
end.

